chef/cookbooks/nova/recipes/ceph.rb

```diff
@@ -0,0 +1,185 @@
+    block do
+      # Check if libvirt was installed
+      virsh_hostname = Mixlib::ShellOut.new("virsh hostname")
+      virsh_hostname.run_command
+      virsh_hostname.error!
+
+      # First remove conflicting secrets due to same usage name
+      virsh_secret = Mixlib::ShellOut.new("virsh secret-list")
+      secret_list = virsh_secret.run_command.stdout
+      virsh_secret.error!
+
+      secret_lines = secret_list.strip.split("\n")
+      if secret_lines.length < 2 || !secret_lines[0].start_with?("UUID") || !secret_lines[1].start_with?("----")
+        raise "cannot fetch list of libvirt secret"
+      end
+      secret_lines.shift(2)
+
+      secret_lines.each do |secret_line|
+        secret_uuid = secret_line.split(" ")[0]
+        cmd = ["virsh", "secret-dumpxml", secret_uuid]
+        virsh_secret_dumpxml = Mixlib::ShellOut.new(cmd)
+        secret_xml = virsh_secret_dumpxml.run_command.stdout
+        virsh_secret_dumpxml.error!
+        # some secrets might not be ceph-related, skip these
+        next if secret_xml.index("<usage type='ceph'>").nil?
+
+        # lazy xml parsing
+        re_match = %r[<usage type='ceph'>.*<name>(.*)</name>]m.match(secret_xml)
+        next if re_match.nil?
+        secret_usage = re_match[1]
+        undefine = false
+
+        if secret_uuid == rbd_uuid
+           undefine = true if secret_usage != "client.#{rbd_user} secret"
+        else
+          undefine = true if secret_usage == "client.#{rbd_user} secret"
+        end
+
+        if undefine
+          cmd = ["virsh", "secret-undefine", secret_uuid]
+          virsh_secret_undefine = Mixlib::ShellOut.new(cmd)
+          virsh_secret_undefine.run_command
+          virsh_secret_undefine.error!
+        end
+      end
+
+      # Now add our secret and its value
+      cmd = ["ceph", "auth", "get-key", "client.#{rbd_user}" ]
+      ceph_get_key = Mixlib::ShellOut.new(cmd)
+      client_key = ceph_get_key.run_command.stdout
+      ceph_get_key.error!
+
+      cmd = ["virsh", "secret-get-value", rbd_uuid ]
+      virsh_secret_get_value = Mixlib::ShellOut.new(cmd)
+      secret = virsh_secret_get_value.run_command.stdout.chomp.strip
+      
+      if secret != client_key    
+        # It seems that virsh secret-define returns '' instead of 
+        cmd = ["virsh", "secret-define", "--file", secret_file_path]
+        virsh_secret_define = Mixlib::ShellOut.new(cmd)
+        secret_uuid_out = virsh_secret_define.run_command.stdout
+        
+        if secret_uuid_out.scan(/(\S{8}-\S{4}-\S{4}-\S{4}-\S{12})/)
+          cmd = ["virsh", "secret-set-value", "--secret", rbd_uuid, "--base64", client_key]
+          virsh_secret_set_value = Mixlib::ShellOut.new(cmd)
+          virsh_secret_set_value.run_command
+          virsh_secret_set_value.error!
+        else
+          raise "Libvirt secret for UUID #{rbd_uuid} was not created properly."
+        end
+      end
+
+      cmd = ["rm", "-f", secret_file_path]
```

Comment:
I'm not sure, if this will work in `ruby_block`, but I will check it. Previously I was trying to use file resource but it failed.